### PR TITLE
(FM-8465) Document architecture

### PR DESCRIPTION
**Pull request by michaeltlombardi**

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -1,4 +1,73 @@
 # PowerShell Manager: Design and Architecture
 
-<!-- TODO: Fill this out! -->
-<!-- NOTE: Not multi-thread safe, executions against the same manager in parallel can get _weird_ -->
\ No newline at end of file
+<!-- NOTE: Not multi-thread safe, executions against the same manager in parallel can get _weird_ -->
```
Comment (michaeltlombardi) [inline DESIGN.md]: TODO: Remove this.

File: DESIGN.md [hunk 2]
```diff
@@ -1,4 +1,73 @@
...
+
+Managing Windows infrastructure from ruby is difficult - much of the interaction with the tools and APIs is performed via PowerShell, and PowerShell remains the most common language for operators of Windows machines to know.
```
Comment (glennsarti) [inline DESIGN.md]: ruby or Ruby?

File: DESIGN.md [hunk 3]
```diff
@@ -1,4 +1,73 @@
...
+Calling PowerShell from ruby has a few problems, but the most painful one is speed - shelling out requires spinning up the PowerShell process, executing a single pipeline, and tearing it back down.
+
+This gem allows the use of a long-lived manager to which ruby can send PowerShell invocations and receive the exection output.
+This reduces the time to execute a second command from seconds to milliseconds.
```
Comment (glennsarti) [inline DESIGN.md]: ` execute a second command from seconds to milliseconds.`

Perhaps reword.  second vs seconds was initially confusing.  Perhaps "subsequent" or "next"

File: DESIGN.md [hunk 4]
```diff
@@ -1,4 +1,73 @@
...
+
+The manager operates by instantiating a custom PowerShell host process to which ruby can then send commands over a pipe--
```
Comment (glennsarti) [inline DESIGN.md]: `over a pipe-`  Perhaps over `an IO pipe`

File: DESIGN.md [hunk 5]
```diff
@@ -1,4 +1,73 @@
...
+on Windows machines, named pipes, on Unix/Linux, Unix sockets.
```
Comment (glennsarti) [inline DESIGN.md]: nit : Unix Domain Sockets

File: DESIGN.md [hunk 6]
```diff
@@ -1,4 +1,73 @@
...
+
+## Communication Between Ruby and PowerShell Host Process
+
+Communication between ruby and the PowerShell host process uses binary encoded strings with a 4-byte prefix indicating how long the message is.
```
Comment (glennsarti) [inline DESIGN.md]: Perhaps link to what a TLV is? https://en.wikipedia.org/wiki/Type-length-value

**Comment by glennsarti** [comment]
Could really do with diagrams (ASCII art?)

File: DESIGN.md (hunk 2 above)
Comment (jpogran) [reply DESIGN.md]: If you refer to PowerShell as cased, then Ruby gets the same treatement

File: DESIGN.md (hunk 3 above)
Comment (jpogran) [reply DESIGN.md]: To further clarify, we should add something along the lines of 'this reduces the time to execute *the* command, not the total time it takes the command to execute.'

File: DESIGN.md (hunk 2 above)
Comment (jpogran) [reply DESIGN.md]: We shouldn't start the document with a negative, and it's not the point of this gem to manage windows from ruby, its to make using powershell easier from ruby. We don't have to justify using powershell on windows, we have to justify using our gem.

This gem reduces the difficulty of running PowerShell in Ruby by X, Y, and Z.

Comment (michaeltlombardi) [reply DESIGN.md]: Resolved with a substantial reword, I think.

File: DESIGN.md (hunk 3 above)
Comment (michaeltlombardi) [reply DESIGN.md]: Resolved with a substantial reword, I think.

**Comment by michaeltlombardi** [comment]
So, GitHub is goofy and won't show the image until it's committed... here's the [link to view it](https://mermaidjs.github.io/mermaid-live-editor/#/view/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5wYXJ0aWNpcGFudCBSdWJ5XG5wYXJ0aWNpcGFudCBQb3dlclNoZWxsIE1hbmFnZXJcblJ1YnktLT5SdWJ5OiBleGVjdXRlKCdTb21lIGNvbW1hbmQnKVxuUnVieS0tPlJ1Ynk6IFdyYXAgJ1NvbWUgQ29tbWFuZCcgaW4gc3RhbmRhcmRpemVkIHNjcmlwdCBibG9ja1xuUnVieS0-PlBvd2VyU2hlbGwgTWFuYWdlcjogJ1NvbWUgQ29tbWFuZCcgaW4gc2NyaXB0IGJsb2NrIGFzIGJpbmFyeS1lbmNvZGVkIHN0cmluZyB3aXRoIGxlbmd0aCBwcmVmaXhcblBvd2VyU2hlbGwgTWFuYWdlci0tPj5Qb3dlclNoZWxsIE1hbmFnZXI6IEV4ZWN1dGUgJ1NvbWUgQ29tbWFuZCdcblBvd2VyU2hlbGwgTWFuYWdlci0-PlJ1Ynk6IFJlc3VsdCBhcyBiaW5hcnktZW5jb2RlZCBzdHJpbmcgd2l0aCBsZW5ndGggcHJlZml4IiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifX0).

**Comment by glennsarti** [comment]
Good start

**Comment by michaeltlombardi** [comment]
@glennsarti What other architecture concerns or concepts would you like to see documented/diagram'd?

**Comment by glennsarti** [comment]
Not sure yet...  Probably how to use the classes.

**Pull request merged**